### PR TITLE
[IMP/REF] web, website: migrate some of the frontend code to use wowl infrastructure

### PR DESCRIPTION
**Pull request by sdegueldre**
See also https://github.com/odoo/enterprise/pull/19258

### CODE REVIEW

File: addons/web/static/src/legacy/js/public/public_widget.js [hunk 1]
```diff
@@ -110,6 +106,7 @@ var RootWidget = Widget.extend({
      * @param {OdooEvent} ev
      */
     _onGetSession: function (event) {
+        console.log('in _onGetSession from public_widget');
```
Comment (ged-odoo) [inline addons/web/static/src/legacy/js/public/public_widget.js:109]: console.log

File: addons/web/__manifest__.py [hunk 2]
```diff
@@ -196,14 +196,38 @@
 
             ('include', 'web._assets_bootstrap'),
 
+            'web/static/src/env.js',
```
Comment (ged-odoo) [inline addons/web/__manifest__.py:200]: haha, it looks like you are getting to what I predicted: 1 bundle for frontend and 1 bundle for backend

File: addons/web/static/src/legacy/js/public/public_root_instance.js [hunk 3]
```diff
@@ -15,19 +15,52 @@ import {
...
-const { Component } = owl;
+const { Component, tags, mount } = owl;
 
 /**
  * Configure Owl with the public env
  */
 owl.config.mode = legacyEnv.isDebug() ? "dev" : "prod";
 owl.Component.env = legacyEnv;
 
+class MainComponentsContainer extends Component {
```
Comment (ged-odoo) [inline addons/web/static/src/legacy/js/public/public_root_instance.js]: i wonder if it would make sense to create that component in webclient, so we do not duplicate code here

File: addons/web/static/src/core/assets.js [hunk 4]
```diff
@@ -99,7 +99,7 @@ const loadCSS = memoize(function loadCSS(url) {
  * @returns {Promise<XMLDocument>} A Promise of an XML document containing the
  *      owl templates.
  */
-const loadBundleTemplates = memoize(async function loadBundleTemplates(name) {
+export const loadBundleTemplates = memoize(async function loadBundleTemplates(name) {
```
Comment (ged-odoo) [inline addons/web/static/src/core/assets.js:102]: can't you use loadBundle instead?

File: addons/web/static/src/legacy/js/public/public_root_instance.js [hunk 5]
```diff
@@ -43,14 +76,17 @@ export default lazyloader.allScriptsLoaded.then(async () => {
     serviceRegistry.add("legacy_notification", makeLegacyNotificationService(legacyEnv));
     serviceRegistry.add("legacy_crash_manager", makeLegacyCrashManagerService(legacyEnv));
     serviceRegistry.add("legacy_dialog_mapping", makeLegacyDialogMappingService(legacyEnv));
-
     await Promise.all([owl.utils.whenReady(), session.is_bound]);
-    //legacyEnv.qweb.addTemplates(session.owlTemplates);
 
     const wowlEnv = makeEnv();
+    wowlEnv.qweb.addTemplates(await loadBundleTemplates("web.assets_frontend"));
```
Comment (ged-odoo) [inline addons/web/static/src/legacy/js/public/public_root_instance.js]: these templates were not loaded before?

File: addons/web/static/src/legacy/js/public/public_root_instance.js [hunk 6]
```diff
@@ -43,14 +76,17 @@ export default lazyloader.allScriptsLoaded.then(async () => {
...
     await startServices(wowlEnv);
 
-    const adapter = new ComponentAdapter(null, { Component });
+    await mount(MainComponentsContainer, {
```
Comment (ged-odoo) [inline addons/web/static/src/legacy/js/public/public_root_instance.js]: note that this is done sequentially before the componentadapter for the publicroot widget... maybe worth it to parallelize it ?

File: addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js [hunk 7]
```diff
@@ -394,7 +393,13 @@ odoo.define('partner_autocomplete.tests', function (require) {
                 '<field name="name" widget="field_partner_autocomplete"/>' +
                 '</form>',
             services: {
-                notification: NotificationService,
+                notification: {
+                    notify(notification) {
+                        debugger;
```
Comment (ged-odoo) [inline addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js:398]: debugger

File: addons/web/static/src/core/l10n/localization_service.js [hunk 8]
```diff
@@ -6,23 +6,37 @@ import { strftimeToLuxonFormat } from "./dates";
 import { localization } from "./localization";
 import { translatedTerms, _t } from "./translation";
 
+const defaultLangParameters = {
```
Comment (ged-odoo) [inline addons/web/static/src/core/l10n/localization_service.js:9]: it's not obvious why we need default values. if this commit is merged, we need some explanations. is it for the frontend?

File: addons/web/static/src/legacy/js/public/public_root_instance.js (hunk 5 above)
Comment (sdegueldre) [reply addons/web/static/src/legacy/js/public/public_root_instance.js]: There was no templates in this bundle before. Now it contains the wowl notifications template, and will probably contain more things in the future

File: addons/web/static/src/core/l10n/localization_service.js (hunk 8 above)
Comment (sdegueldre) [reply addons/web/static/src/core/l10n/localization_service.js:9]: It's for the login because there is no session_info in the login layout, but maybe I should add the frontend_session_info to the login layout instead

File: addons/web/__manifest__.py (hunk 2 above)
Comment (sdegueldre) [reply addons/web/__manifest__.py:200]: I'm trying this approach because you suggested it, self fulfilling prophecy?

File: addons/web/static/src/legacy/js/public/public_root_instance.js (hunk 3 above)
Comment (sdegueldre) [reply addons/web/static/src/legacy/js/public/public_root_instance.js]: Oh yeah, I intended to do that, I was just trying this out to see if it would work and forgot to factor it out of the webclient template and just use this component instead.

File: addons/web/static/src/core/assets.js (hunk 4 above)
Comment (sdegueldre) [reply addons/web/static/src/core/assets.js:102]: Let's talk about the API on monday. Not a fan of using a very generic API like loadBundle when you only want to load the templates, although in this case we would like to load the whole bundle, but loadBundle currently only allows to load templates (but with a clunkier API)

File: addons/web/static/src/legacy/js/public/public_root_instance.js (hunk 6 above)
Comment (sdegueldre) [reply addons/web/static/src/legacy/js/public/public_root_instance.js]: Neither of these tasks is IO-bound, parallelizing them doesn't bring any performance benefit and makes the code harder to read. We also want the `MainComponentsContainer` to be ready before we call `publicRoot.attachTo` as it may call services that rely on the `MainComponentsContainer` while attaching (eg, creating a notification in start)

File: addons/web/static/src/core/l10n/localization_service.js [hunk 9]
```diff
@@ -6,22 +6,37 @@ import { strftimeToLuxonFormat } from "./dates";
...
+    "grouping": "[]",
+    "decimal_point": ".",
+    "thousands_sep": ",",
+};
+
 export const localizationService = {
     dependencies: ["user"],
     start: async (env, { user }) => {
-        const cacheHashes = odoo.session_info.cache_hashes;
+        const cacheHashes = odoo.session_info && odoo.session_info.cache_hashes || {};
```
Comment (aab-odoo) [inline addons/web/static/src/core/l10n/localization_service.js:21]: `const sessionInfo = odoo.session_info || {};`

File: addons/web/static/src/legacy/js/public/public_root.js [hunk 10]
```diff
@@ -135,7 +151,7 @@ var PublicRoot = publicWidget.RootWidget.extend({
      * @override
      */
     _getRegistry: function () {
-        return publicRootRegistry;
+        return registry.category("public_root_widgets");
```
Comment (aab-odoo) [inline addons/web/static/src/legacy/js/public/public_root.js:154]: mix legacy widgets and wowl registries?

File: addons/http_routing/controllers/main.py [hunk 11]
```diff
@@ -9,7 +9,7 @@
 class Routing(Home):
 
     @http.route('/website/translations/<string:unique>', type='http', auth="public", website=True)
-    def get_website_translations(self, unique, lang, mods=None):
+    def get_website_translations(self, unique, lang=None, mods=None):
```
Comment (aab-odoo) [inline addons/http_routing/controllers/main.py:12]: beware of commit message ([REF] web, website)

Comment (aab-odoo) [reply addons/http_routing/controllers/main.py:12]: also, I'll trust you on this, no idea why you need to change

File: addons/web/static/src/core/main_components_container.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    <t t-foreach="Components" t-as="C" t-key="C[0]">
+        <NotUpdatable>
+            <ErrorHandler onError="error => handleComponentError(error, C)">
+                <t t-component="C[1].Component" t-props="C[1].props"/>
+            </ErrorHandler>
+        </NotUpdatable>
+    </t>
+</div>
+`;
+MainComponentsContainer.components = { NotUpdatable, ErrorHandler };
```
Comment (aab-odoo) [inline addons/web/static/src/core/main_components_container.js:37]: enable tooling!!!! :-)

File: addons/test_website/static/tests/tours/error_views.js [hunk 13]
```diff
@@ -49,8 +49,8 @@ tour.register('test_error_website', {
         trigger: 'a[href="/test_internal_error_json"]',
     }, {
         content: "rpc error 500 modal is an ErrorDialog",
-        extra_trigger: 'div.o_dialog_error.modal-body div.alert.alert-warning',
-        trigger: 'button.btn.btn-primary[type="button"]',
+        extra_trigger: 'div.o_dialog_error.modal-content div.alert.alert-warning',
+        trigger: '.modal-footer button.btn.btn-primary',
```
Comment (aab-odoo) [inline addons/test_website/static/tests/tours/error_views.js:53]: commit message [REF] web, website, website_forum

File: addons/web/static/src/frontend/error_notifications.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
+/** @odoo-module */
```
Comment (aab-odoo) [inline addons/web/static/src/frontend/error_notifications.js:1]: should remain in web/(...)/frontend/ ?

Comment (aab-odoo) [reply addons/web/static/src/frontend/error_notifications.js:1]: I think we talked about having a public/ folder (for the database_manager at least)

**Review by aab-odoo — APPROVED** [review]

File: addons/web/static/src/legacy/js/public/public_widget.js [hunk 15]
```diff
@@ -19,18 +17,12 @@ import Widget from 'web.Widget';
  * @todo Merge with 'PublicWidget' ?
  */
 var RootWidget = Widget.extend({
-    custom_events: _.extend({}, Widget.prototype.custom_events || {}, {
-        'registry_update': '_onRegistryUpdate',
-        'get_session': '_onGetSession',
-    }),
```
Comment (qsm-odoo) [inline addons/web/static/src/legacy/js/public/public_widget.js:26]: Removing this but keeping the `_onRegistryUpdate`, same for `_onGetSession`... I don't follow this commit :grimacing: And what about the code using the `get_session` event? :thinking:

File: addons/http_routing/controllers/main.py (hunk 11 above)
Comment (qsm-odoo) [reply addons/http_routing/controllers/main.py:12]: I am a bit more curious, why that change? :wink:

File: addons/web/__manifest__.py (hunk 2 above)
Comment (qsm-odoo) [reply addons/web/__manifest__.py:200]: Actually you warned me about this but reading this conversation I wonder why not in common? :thinking: I guess too much "common" places which don't need that? Not sure, is not it worth to at least add those in a common template which is 'include' in both assets?

File: addons/web/static/src/core/main_components_container.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
+/** @odoo-module */
+import { registry } from "./registry";
+import { NotUpdatable, ErrorHandler } from "./utils/components";
+
+const { Component, tags } = owl;
+
+export class MainComponentsContainer extends Component {
```
Comment (qsm-odoo) [inline addons/web/static/src/core/main_components_container.js:7]: Could this be on its own commit too maybe? Hard to follow the direct link with frontend convertion

File: addons/web/static/src/core/user_service.js [hunk 17]
```diff
@@ -55,15 +55,15 @@ export const userService = {
             },
             get db() {
                 const res = {
-                    name: sessionInfo.db,
```
Comment (qsm-odoo) [inline addons/web/static/src/core/user_service.js:58]: Is this a new required convention? The commit is kinda polluted with those changes and I wonder if your editor fixed them all by mistake or something (the opposite convention was decided a long time ago :grimacing:)

File: addons/web/static/src/legacy/js/public/public_root.js [hunk 18]
```diff
@@ -328,6 +348,46 @@ var PublicRoot = publicWidget.RootWidget.extend({
...
+    const publicRoot = new RootWidget(adapter);
+    await Promise.all([
+        mount(MainComponentsContainer, {
+            target: document.body,
+            env: wowlEnv,
+        }),
+        publicRoot.attachTo(document.body),
+    ]);
+    return publicRoot;
+}
```
Comment (qsm-odoo) [inline addons/web/static/src/legacy/js/public/public_root.js]: Ok, kinda lost in all of this but I guess it's because I don't know OWL (and unfortunately no time to check it right now). Kinda curious why the "root" container... must have a container itself? :thinking: Can't it *be* that container somehow?

**Review by qsm-odoo — COMMENTED** [review]
I'll need to test this but it seems good, hopefully this step by step approach will lead somewhere :+1: Out of curiosity, did you compute the added bundle size of adding all services etc?

Here are some comments :+1: (some more important than others)

File: addons/web/static/src/core/errors/error_dialogs.js [hunk 19]
```diff
@@ -25,7 +25,7 @@ export class ErrorDialog extends Dialog {
     setup() {
         super.setup();
         this.state = useState({
-            showTraceback: false,
+            showTraceback: false
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web/static/src/core/errors/error_dialogs.js:34]: Yup, we are now using prettier and apparently you can't configure it to have trailing commas only when multiline, so it's that or have trailing commas even on single line objects. And it's not just linted, but prettified automatically in a pre-commit hook that we're required to run on all files in web that aren't in `/legacy`. I don't like this either but I've voiced my opinion more than once and I've got bigger fish to fry than to continue arguing about this. I'll pay particular attention to not let this bleed into files in the website module.

File: addons/web/static/src/legacy/js/public/public_root.js [hunk 20]
```diff
@@ -6,8 +6,7 @@ import env from 'web.public_env';
 import session from 'web.session';
 import utils from 'web.utils';
 import publicWidget from 'web.public.widget';
-
-var publicRootRegistry = new publicWidget.RootWidgetRegistry();
+import { registry } from '../../../core/registry';
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web/static/src/legacy/js/public/public_root.js:10]: The idea is generally to try to use relative imports when in the same module, because that allows editor tooling like VSCode to find the file and have auto-completion without having to define resolve paths explicitly in your tsconfig/jsconfig, but this convention is followed quite loosely, and in this case I agree that the path makes more sense when reading if it's an absolute path so I'll change it.

File: addons/web/static/src/legacy/js/public/public_widget.js (hunk 15 above)
Comment (sdegueldre) [reply addons/web/static/src/legacy/js/public/public_widget.js:26]: Whoops, need to remove those dead methods. `get_session` event is now handled by `ComponentAdapter` that I've set as this widget's parent, it does a bunch of `trigger_up` and `call_service` translation.

File: addons/http_routing/controllers/main.py (hunk 11 above)
Comment (sdegueldre) [reply addons/http_routing/controllers/main.py:12]: Because the `WebClient` method which we call has this param marked as optional, so in the localization service, when we call the translation route, if the lang is not present, we don't put the param in the URL, whereas in the previous code for the frontend, we would always add it to the URL, and if not present it would be an empty string. Instead of trying to share a translation object globally through some dirty tricks (like setting a promise on the global odoo object), we decided to leverage browser caching, so the URLs have to be the same, and I thought it made more sense to make the parameter optional (functionally, it already is, it does the same thing when you pass an empty string as when you don't pass it at all) rather than always putting the lang param in the URL even if it's empty.

File: addons/web/static/src/core/user_service.js (hunk 17 above)
Comment (sdegueldre) [reply addons/web/static/src/core/user_service.js:58]: (see above comment about us now running prettier on all files that have been touched in web in pre-commit hook)

File: addons/web/static/src/legacy/js/public/public_root.js (hunk 18 above)
Comment (sdegueldre) [reply addons/web/static/src/legacy/js/public/public_root.js]: Not sure what you mean here. The publicRoot widget doesn't have a container (though it has a parent to handle trigger_up but that parent is not in the DOM).

File: addons/web/__manifest__.py (hunk 2 above)
Comment (sdegueldre) [reply addons/web/__manifest__.py:200]: @ged-odoo suggested that we stop having a common bundle altogether to make asset management easier: in web we need all of core so we can just use one glob (`/wev/static/src/core/**/*.js`) which makes it easy to see at a glance that we use everything, whereas in frontend we really want to pick and choose only what we need to reduce the bundle size. I think the end goal was to completely avoid having a "common" bundle and simply have one bundle per use case, Géry might be able to tell you more about the exact reasoning. I'm open to moving most of this to common though, since most of this is in fact common and needed to have the basics working (error service, RPC, notificaitons)

File: addons/web/static/src/core/main_components_container.js (hunk 16 above)
Comment (sdegueldre) [reply addons/web/static/src/core/main_components_container.js:7]: Sure, I'll move that to its own commit for clarity. Basically I just factored out something that the webclient was already doing so that I could reuse it, but I agree that it's sufficiently self-contained that it can be in its own commit.

**Comment by sdegueldre** [comment]
> Out of curiosity, did you compute the added bundle size of adding all services etc?

Looks like `assets_frontend_lazy` went from ~170kb to ~220kb, though this is still bound to drop after I remove some of the other things that are currently still included in them but for which equivalents now exist in the new code (like the KeyboardNavigationMixin, RainbowMan and DebugManager), after that we should be close to parity as far as bundle size is concerned hopefully.

File: addons/web/static/src/legacy/js/public/public_root.js (hunk 18 above)
Comment (qsm-odoo) [reply addons/web/static/src/legacy/js/public/public_root.js]: container/parent: my point ons that "my root stuff has a parent" = it seem strange, but ok I guess I can live with "root" "of the DOM" but not of the objects hierarchy

File: addons/web/__manifest__.py (hunk 2 above)
Comment (qsm-odoo) [reply addons/web/__manifest__.py:200]: Same advice on the **assets** point of view (like there is no real value of having a common bundle including stuff that is loaded separately and that we are never too sure we need entirely). But picking stuff in core seems unperfect IMHO = you have to know what core module need each other and, also, what is in "core" should be available everywhere even if unused IMHO (if it is something very specific, it should not be in core).

In the end: maybe just a matter of splitting the framework "core" in "core-for-backend" and "core-for-odoo"? (not with those names obviously)

File: addons/web/static/src/legacy/js/public/public_root.js (hunk 18 above)
Comment (sdegueldre) [reply addons/web/static/src/legacy/js/public/public_root.js]: Right. This is not a permanent solution, the next step will be converting the publicRoot/websiteRoot to fully fledged owl components, and in that scenario they won't need to have the ComponentAdapter as a parent anymore, they will be the actual root in both the DOM and in the component hierarchy, and then I'll work my way down to the descendants.

**Comment by sdegueldre** [comment]
@qsm-odoo I think I addressed the points you raised, can we move forward with this, anything else still sticking out to you?

**Comment by qsm-odoo** [comment]
@sdegueldre Except for my comment here but I guess you can keep that in mind? https://github.com/odoo/odoo/pull/72675#discussion_r668566845

But sure, go ahead, I tested a bit notification, crashmanager and other stuff, seems good :+1: Don't really have time to check more than I did anyway :grimacing:

**Comment by sdegueldre** [comment]
@robodoo rebase-ff

**Review by aab-odoo — APPROVED** [review]

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo rebase-ff
